src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java

```diff
@@ -408,24 +426,47 @@ FireStoreDirectoryEntry computeDirectory(
         String fullPath = fireStoreUtils.getFullPath(dirEntry.getPath(), dirEntry.getName());
         List<FireStoreDirectoryEntry> enumDir = directoryDao.enumerateDirectory(firestore, snapshotId, fullPath);
 
-        // Recurse to compute results from underlying directories
         List<FireStoreDirectoryEntry> enumComputed = new ArrayList<>();
-        for (FireStoreDirectoryEntry dirItem : enumDir) {
-            if (dirItem.getIsFileRef()) {
-                // Read the file metadata to get the size and checksum. We do a bit of a hack and copy
-                // the size and checksums into the in-memory dirItem. That way we only compute on the directory
-                // objects.
-                FireStoreFile file =
-                    fileDao.retrieveFileMetadata(firestore, dirItem.getDatasetId(), dirItem.getFileId());
-
-                dirItem
-                    .size(file.getSize())
-                    .checksumMd5(file.getChecksumMd5())
-                    .checksumCrc32c(file.getChecksumCrc32c());
-
-                enumComputed.add(dirItem);
-            } else {
-                enumComputed.add(computeDirectory(firestore, snapshotId, dirItem, updateBatch));
+
+        // Recurse to compute results from underlying directories
+        try (Stream<FireStoreDirectoryEntry> stream = enumDir.stream()) {
+            enumComputed.addAll(stream
+                .filter(f -> !f.getIsFileRef())
+                .map(f -> {
+                    try {
+                        return computeDirectory(firestore, snapshotId, f, updateBatch);
+                    } catch (InterruptedException e) {
+                        throw new DirectoryMetadataComputeException("Error computing directory metadata", e);
+                    }
+                })
+                .collect(Collectors.toList())
+            );
+        }
+
+        // Collect metadata for file objects in the directory
+        try (Stream<FireStoreDirectoryEntry> stream = enumDir.stream()) {
+            // Group FireStoreDirectoryEntry objects by dataset Id to process one dataset at a time
+            final Map<String, List<FireStoreDirectoryEntry>> fileRefsByDatasetId = stream
+                .filter(FireStoreDirectoryEntry::getIsFileRef)
+                .collect(Collectors.groupingBy(FireStoreDirectoryEntry::getDatasetId));
+
+            for (Map.Entry<String, List<FireStoreDirectoryEntry>> entry : fileRefsByDatasetId.entrySet()) {
+                // Retrieve the file metadata from Firestore
+                final List<FireStoreFile> fireStoreFiles =
+                    fileDao.batchRetrieveFileMetadata(firestore, entry.getKey(), entry.getValue());
+
+                final AtomicInteger index = new AtomicInteger(0);
```

Comment:
Why is atomic needed here?

Reply:
Ah, that's because since I'm using index in a lambda, it needs to be effectively final.  I tend to like using `AtomicIntegers` for this.  I could have some something like storing it in a one element integer array I just fund this easier to work with